internal/target/service_list_refresh.go

```diff
@@ -38,25 +41,32 @@ func ListRefresh(
 	if tok == nil {
 		return nil, errors.New(ctx, errors.InvalidParameter, op, "missing token")
 	}
+	rt, ok := tok.Subtype.(*listtoken.StartRefreshToken)
+	if !ok {
+		return nil, errors.New(ctx, errors.InvalidParameter, op, "token did not have a start-refresh token component")
+	}
 	if repo == nil {
 		return nil, errors.New(ctx, errors.InvalidParameter, op, "missing repo")
 	}
 
-	listItemsFn := func(ctx context.Context, tok *refreshtoken.Token, lastPageItem Target, limit int) ([]Target, error) {
+	listItemsFn := func(ctx context.Context, lastPageItem Target, limit int) ([]Target, error) {
 		opts := []Option{
 			WithLimit(limit),
 		}
 		if lastPageItem != nil {
 			opts = append(opts,
 				WithStartPageAfterItem(lastPageItem),
 			)
-		} else {
-			opts = append(opts,
-				WithStartPageAfterItem(tok.LastItem()),
-			)
 		}
-		return repo.listTargets(ctx, opts...)
+		// Add the database read timeout to account for any creations missed due to concurrent
+		// transactions in the original list pagination phase.
+		return repo.listTargetsRefresh(ctx, rt.PhaseEndTime.Add(-globals.DatabaseReadTimeout), opts...)
+	}
+	listDeletedIdsFn := func(ctx context.Context, since time.Time) ([]string, error) {
+		// Add the database read timeout to account for any deletions missed due to concurrent
+		// transactions in previous requests.
+		return repo.listDeletedIds(ctx, since.Add(-globals.DatabaseReadTimeout))
```

Comment:
This seemed like the right place to do the database read timeout buffer compensation, rather than in the pagination or listtoken packages.

Reply:
I'm happy to change it, but it seems like the knowledge of this subtraction belongs in the specific domain package rather than in the listtoken domain. The listtoken, to me, is responsible for providing a type that maintains the state of the pagination and helps transition between the phases. It doesn't need to (shouldn't?) know about the database specific details such as this timestamp adjustment. It's also only used in the refresh phase, so if we put this in the listtoken, how do we make clear when the timestamp has been adjusted already? It seems simpler to me to handle that at query time than that token construction time.

Reply:
I think that makes sense. I think the downside is that there are a lot of places where this subtraction needs to be done.

Reply:
It's true, but at least it's a simple operation, and I intend to include this comment in each place to add context to the reader.